From e6578fdd4df9f725dd53442522b8263363d5d46c Mon Sep 17 00:00:00 2001
From: Hou Qi <qi.hou@nxp.com>
Date: Wed, 27 Sep 2023 15:29:30 +0800
Subject: [PATCH 54/75] v4l2: add downscale support for imx95 wave6 vpu dec

Wave6 driver sets mimimum downscale size to align up with 4,
and sets downscale size to align up with 4 if selection flag
is set as V4L2_SEL_FLAG_GE.

For a 3840x2160 stream, if caps require 1/8th downscale size
480x270, driver reports 480x272. There will be a crop 2 at bottom.

Upstream-Status: Pending
Signed-off-by: Xiaolei Wang <xiaolei.wang@windriver.com>
---
 sys/v4l2/gstv4l2object.c   | 21 ++++++++-----
 sys/v4l2/gstv4l2object.h   |  4 ++-
 sys/v4l2/gstv4l2videodec.c | 64 ++++++++++++++++++++++++++++++++++++++
 3 files changed, 81 insertions(+), 8 deletions(-)

diff --git a/sys/v4l2/gstv4l2object.c b/sys/v4l2/gstv4l2object.c
index ef825de..8eb1ee7 100644
--- a/sys/v4l2/gstv4l2object.c
+++ b/sys/v4l2/gstv4l2object.c
@@ -581,6 +581,8 @@ gst_v4l2_object_new (GstElement * element,
 
   v4l2object->poll = gst_poll_new (TRUE);
   v4l2object->can_poll_device = TRUE;
+  v4l2object->crop_width = 0;
+  v4l2object->crop_height = 0;
 
   /* We now disable libv4l2 by default, but have an env to enable it. */
 #ifdef HAVE_LIBV4L2
@@ -3395,7 +3397,7 @@ gst_v4l2_object_reset_compose_region (GstV4l2Object * obj)
   GST_V4L2_CHECK_OPEN (obj);
 
   sel.type = obj->type;
-  sel.target = V4L2_SEL_TGT_COMPOSE_DEFAULT;
+  sel.target = V4L2_SEL_TGT_COMPOSE;
 
   if (obj->ioctl (obj->video_fd, VIDIOC_G_SELECTION, &sel) < 0) {
     if (errno == ENOTTY) {
@@ -4622,7 +4624,7 @@ gst_v4l2_object_acquire_format (GstV4l2Object * v4l2object, GstVideoInfo * info)
   /* Use the default compose rectangle */
   memset (&sel, 0, sizeof (struct v4l2_selection));
   sel.type = v4l2object->type;
-  sel.target = V4L2_SEL_TGT_COMPOSE_DEFAULT;
+  sel.target = V4L2_SEL_TGT_COMPOSE;
   if (v4l2object->ioctl (v4l2object->video_fd, VIDIOC_G_SELECTION, &sel) >= 0) {
     r = &sel.r;
   } else {
@@ -4635,10 +4637,15 @@ gst_v4l2_object_acquire_format (GstV4l2Object * v4l2object, GstVideoInfo * info)
   if (r) {
     align.padding_left = r->left;
     align.padding_top = r->top;
-    align.padding_right = width - r->width - r->left;
-    align.padding_bottom = height - r->height - r->top;
-    width = r->width;
-    height = r->height;
+    align.padding_right =
+        (gint) width - (gint) r->width - r->left - v4l2object->crop_width;
+    align.padding_bottom =
+        (gint) height - (gint) r->height - r->top - v4l2object->crop_height;
+    width = (gint) r->width + v4l2object->crop_width;
+    height = (gint) r->height + v4l2object->crop_height;
+    GST_INFO_OBJECT (v4l2object->dbg_obj,
+        "caps size: %ux%u, padding size: %ux%u", width, height,
+        align.padding_right, align.padding_bottom);
   }
 
   switch (fmt.fmt.pix.field) {
@@ -5537,7 +5544,7 @@ gst_v4l2_object_decide_allocation (GstV4l2Object * obj, GstQuery * query)
 
   /* aovid copy Amphion tiled frame buffer for un-active video track */
   /* also to avoid copy Hantro frame buffer when link v4l2 decoder with fakesink */
-  if (obj->is_amphion || obj->is_hantro) {
+  if (obj->is_amphion || obj->is_hantro || IS_IMX95 ()) {
     can_share_own_pool = TRUE;
     if (min < GST_V4L2_MIN_BUFFERS (obj))
       min = GST_V4L2_MIN_BUFFERS (obj);
diff --git a/sys/v4l2/gstv4l2object.h b/sys/v4l2/gstv4l2object.h
index f1ad009..1c9204f 100644
--- a/sys/v4l2/gstv4l2object.h
+++ b/sys/v4l2/gstv4l2object.h
@@ -238,7 +238,9 @@ struct _GstV4l2Object {
   gboolean skip_try_fmt_probes;
   gboolean is_amphion;
   gboolean is_hantro;
-  
+  gint crop_width;
+  gint crop_height;
+
   guint max_width;
   guint max_height;
 };
diff --git a/sys/v4l2/gstv4l2videodec.c b/sys/v4l2/gstv4l2videodec.c
index f499b71..89dd7d8 100644
--- a/sys/v4l2/gstv4l2videodec.c
+++ b/sys/v4l2/gstv4l2videodec.c
@@ -834,6 +834,68 @@ gst_v4l2_video_dec_wait_for_src_ch (GstV4l2VideoDec * self)
   return flowret;
 }
 
+static void
+gst_v4l2_video_dec_set_selection (GstVideoDecoder * decoder)
+{
+  GstV4l2VideoDec *self = GST_V4L2_VIDEO_DEC (decoder);
+  GstV4l2Object *v4l2object = self->v4l2capture;
+  struct v4l2_selection sel = { 0 };
+  GstCaps *filter, *caps;
+  GstStructure *structure;
+  const GValue *value;
+  gint width = 0, height = 0;
+  gint src_width, src_height;
+
+  src_width = self->v4l2output->info.width;
+  src_height = self->v4l2output->info.height;
+
+  // get the downscale width/height from caps if have
+  filter = gst_caps_new_empty_simple ("video/x-raw");
+  caps = gst_pad_peer_query_caps (decoder->srcpad, filter);
+  structure = gst_caps_get_structure (caps, 0);
+  GST_INFO_OBJECT (v4l2object->dbg_obj, "queried caps: %" GST_PTR_FORMAT, caps);
+
+  value = gst_structure_get_value (structure, "width");
+  if (value && G_VALUE_TYPE (value) == G_TYPE_INT)
+    width = g_value_get_int (value);
+
+  value = gst_structure_get_value (structure, "height");
+  if (value && G_VALUE_TYPE (value) == G_TYPE_INT)
+    height = g_value_get_int (value);
+
+  if (width >= src_width / 8 && width <= src_width &&
+      height >= src_height / 8 && height <= src_height) {
+    GST_INFO_OBJECT (v4l2object->dbg_obj, "want to s_selection %dx%d", width,
+        height);
+
+    // set selection with downscale size
+    sel.type = v4l2object->type;
+    sel.target = V4L2_SEL_TGT_COMPOSE;
+    sel.flags = V4L2_SEL_FLAG_GE;
+    sel.r.left = 0;
+    sel.r.top = 0;
+    sel.r.width = width;
+    sel.r.height = height;
+
+    if (v4l2object->ioctl (v4l2object->video_fd, VIDIOC_S_SELECTION, &sel) < 0) {
+      GST_WARNING_OBJECT (v4l2object->dbg_obj,
+          "Failed to set downscale size with VIDIOC_S_SELECTION: %s",
+          g_strerror (errno));
+      goto out;
+    }
+    v4l2object->crop_width = width - (gint) sel.r.width;
+    v4l2object->crop_height = height - (gint) sel.r.height;
+
+    GST_INFO_OBJECT (v4l2object->dbg_obj, "got downscale size %dx%d",
+        sel.r.width, sel.r.height);
+  }
+
+out:
+  gst_caps_unref (filter);
+  gst_caps_unref (caps);
+  return;
+}
+
 static void
 gst_v4l2_video_dec_loop (GstVideoDecoder * decoder)
 {
@@ -854,6 +916,8 @@ gst_v4l2_video_dec_loop (GstVideoDecoder * decoder)
       goto beach;
     }
 
+    gst_v4l2_video_dec_set_selection (decoder);
+
     GST_DEBUG_OBJECT (decoder, "Setup the capture queue");
     if (G_UNLIKELY (!GST_V4L2_IS_ACTIVE (self->v4l2capture))) {
       if (!gst_video_decoder_negotiate (decoder)) {
-- 
2.43.0

